sycl/include/sycl/detail/kernel_desc.hpp

```diff
@@ -151,6 +168,98 @@ template <class KernelNameType> struct KernelInfo {
 };
 #endif //__SYCL_UNNAMED_LAMBDA__
 
+// Built-ins accept an object due to lacking infrastructure support for
+// accepting types. The kernel name type itself isn't used because it might be
+// incomplete, cv-qualified, or not default constructible. Passing an object
+// also allows future extension for SYCL kernels defined as free functions.
+template <typename KNT> struct KernelIdentity {
+  using type = KNT;
+};
+
+template <typename KernelNameType> constexpr unsigned getKernelNumParams() {
+#ifndef __INTEL_SYCL_USE_INTEGRATION_HEADERS
+  return __builtin_sycl_kernel_param_count(KernelIdentity<KernelNameType>());
+#else
+  return KernelInfo<KernelNameType>::getNumParams();
+#endif
+}
+
+template <typename KernelNameType>
+std::vector<kernel_param_desc_t> getKernelParamDescs() {
+  std::vector<kernel_param_desc_t> Result;
+  int NumParams = getKernelNumParams<KernelNameType>();
+  Result.reserve(NumParams);
+  for (int I = 0; I < NumParams; ++I) {
+#ifndef __INTEL_SYCL_USE_INTEGRATION_HEADERS
+    kernel_param_desc_t ParamDesc;
+    ParamDesc.kind =
+        __builtin_sycl_kernel_param_kind(KernelIdentity<KernelNameType>(), I);
+    ParamDesc.info = ParamDesc.kind == kernel_param_kind_t::kind_accessor
+                         ? __builtin_sycl_kernel_param_access_target(
+                               KernelIdentity<KernelNameType>(), I)
+                         : __builtin_sycl_kernel_param_size(
+                               KernelIdentity<KernelNameType>(), I);
+    ParamDesc.offset =
+        __builtin_sycl_kernel_param_offset(KernelIdentity<KernelNameType>(), I);
+    Result.push_back(ParamDesc);
+#else
+    Result.push_back(KernelInfo<KernelNameType>::getParamDesc(I));
+#endif
+  }
+  return Result;
+}
+
+template <typename KernelNameType> constexpr const char *getKernelName() {
+#ifndef __INTEL_SYCL_USE_INTEGRATION_HEADERS
+  return __builtin_sycl_kernel_name(KernelIdentity<KernelNameType>());
+#else
+  return KernelInfo<KernelNameType>::getName();
+#endif
+}
+
+template <typename KernelNameType> constexpr bool isKernelESIMD() {
+  // TODO Needs a builtin counterpart
+  return KernelInfo<KernelNameType>::isESIMD();
```

Comment:
Won't this result in an error now if `isKernelESIMD()` is called in non-integration header mode? Since we don't support ESIMD at the moment in that mode, should we just return false for non-integration header mode?

Reply:
We already do, in both cases. The non-specialized version of KernelInfo defined here has values we will default to (0).

Reply:
But won't you try to access the specialized version when in the new path?

Reply:
Assuming the integration headers are not included, we will default to the non-specialized definition. And assuming they are, I don't see any problems with using the information from there while using the built-ins for everything else.

Reply:
I spoke to Sergey offline. The requested specialization will be implicitly instantiated from the primary template.